google/oauth2/credentials.py

```diff
@@ -71,6 +71,10 @@ def __init__(self, token, refresh_token=None, id_token=None,
                 to obtain authorization. This is a purely informative parameter
                 that can be used by :meth:`has_scopes`. OAuth 2.0 credentials
                 can not request additional scopes after authorization.
+            downscope (bool): Whether to reduce the requested scopes from those
```

Comment:
Hmm. This class is explicitly `ReadOnlyScoped` but `downscope` allows scopes to be modified (albeit not directly), breaking the constraint.

I'd prefer a new method, `credentials.downscope(new_scopes="...")` that explicitly creates a new copy of the credentials with modified scopes. This retains the read-only constraint and makes the downscoping process explicit.

For example:

```
broad_credentials = oauth2.Credentials.from_authorized_user_info(...)
downscoped_credentials = broad_credentials.downscope(["one", "two"])
```

Reply:
The scopes are set at initialization, and changing them afterwords would require changing private state (one or both of _scopes or _downscope). I don't see any way to get around storing that state either, since the refresh function of the new credentials object would need to be aware that it needs to request reduced scopes. It needs to ask for reduced scope every time, since access token request scope defaults to the full scope granted by the refresh token (https://tools.ietf.org/html/rfc6749#section-6). I can't think of any way to implement a downscope function that doesn't circle back around to creating an object that has basically the same implementation as the one proposed here.

Side Note: The credentials as written could already be considered to have mutable scopes, since (per https://tools.ietf.org/html/rfc6749#section-5.1) the server is not required to return an access token with all the scopes of the refresh token.